### PR TITLE
First draft of ROCm support.

### PR DESCRIPTION
**Pull request by mcneish1**
This commit does the simple thing: pass everything through hipify-perl and
compile whatever comes out the other side. Interestingly, that almost works.
Other changes support missing features from the hip runtime, path modifications
required to put everything in the right spot, and differences in hip types.

All of the interesting stuff here is in cmake/HipBuildSystem.cmake.

Known issues:
 * ~~I don't check for the existence of hipify-perl~~
 * Tests and benchmarks are currently disabled
 * ~~Installation likely doesn't work and hasn't been tested~~
 * ~~Header installation works, however installation of the cmake target files isn't quite right~~
 * CMake options given along with enabling ROCm aren't respected or tested
 * ~~Tom was talking about Interesting New Ways of feeding CMake source files
   the other day; this doesn't do any of that~~
 * ~~I'm using add_custom_command wrong: updates to original files aren't seen and~~
   ~~generated files must be deleted for dependency tracking to work~~
 * Both of the casts in src/cuda.cpp are likely wrong in the general case
 * I don't know why in src/cuda.hpp:GPUWait there's both wait_sync and also
   wait_sync_dev_ptr_no_stream_mem_ops, which means what I've done in and around
   src/cuda.cpp and src/helper_kernels.cu may be wrong for reasons other than
   casting and compile guards
 * ~~`set_relative_path` doesn't work for non-ROCm builds (`src/Al.cpp not found`)~~
 * I haven't updated the README yet

### CODE REVIEW

File: src/cuda.hpp [hunk 1]
```diff
@@ -27,12 +27,32 @@
...
+ * Things we use that hip doesn't support:
+ * - cuStreamWaitValue32
+ * - cuStreamWriteValue32
+ * - cuGetErrorString
+ * Of these, the first two are supported by the non-STREAM_MEM_OP code, so we
+ * just need to reimplement cuGetErrorString
+ */
+
+#ifdef AL_HAS_ROCM
+inline hipError_t cuGetErrorString(hipError_t error, const char** pStr)
```
Comment (benson31) [inline src/cuda.hpp]: `error` is unused.

**Comment by benson31** [comment]
The install target is definitely broken.

**Comment by benson31** [comment]
Nonetheless, I was able to build Hydrogen+Aluminum using the build files (with slight modification, e.g., things like OpenMP and references to CUDA).

Another issue is that the MPI-CUDA target (poorly named, but it will be renamed in another PR) needs to be built but there's a CMake logic error that forces CUDA (when we really want ROCm).

File: src/cuda.hpp (hunk 1 above)
Comment (mcneish1) [reply src/cuda.hpp]: Fixed

**Pull request closed without merging**